### PR TITLE
Make global string_container more resilient to different static init orders

### PR DESCRIPTION
**Pull request by reuk**
At the moment, if we initialise a static object which depends on the global string_container, there's no guarantee which object will be initialised first, which means the string_container may be uninitialised when first used. This problem was spotted by @jasigal recently.

This PR wraps the global object in a function which ensures it is initialised before it is first used.

### CODE REVIEW

File: src/util/string_container.cpp [hunk 1]
```diff
@@ -87,3 +85,10 @@ unsigned string_containert::get(const std::string &s)
 
   return r;
 }
+
+/// Get a reference to the global string container.
+string_containert &get_string_container()
+{
+  static string_containert ret{};
```
Comment (tautschnig) [inline src/util/string_container.cpp]: Why are those `{}` required?

**Comment by tautschnig** [comment]
What is the performance difference?

File: src/util/string_container.cpp (hunk 1 above)
Comment (reuk) [inline src/util/string_container.cpp]: They're not, but they make it clear that default-initialisation is intentional.

Comment (tautschnig) [inline src/util/string_container.cpp]: Could be done if done consistently across the code base. Else please remove them for we don't (yet) do it elsewhere.

**Comment by reuk** [comment]
I'm not sure how to measure the performance difference. This PR *shouldn't* affect performance at all, because it only changes the order in which static objects are initialised. The total number of static initialisations is unchanged.

**Comment by tautschnig** [comment]
> I'm not sure how to measure the performance difference.

I believe this is a problem that needs to be addressed before this PR can be merged.

> This PR shouldn't affect performance at all, because it only changes the order in which static objects are initialised. 

How about the non-inlined function calls that will now be necessary each time `dstringt` or `as_string` is invoked?

**Comment by smowton** [comment]
Possible alternative: instead of a `static irep_idt` (if that's what was causing this to begin with),  I've been using a static function local, which gets initialised on first call instead of during library init. For example,

```
const irep_idt &get_some_id() {
  static const irep_idt some_id("some_id");
  return some_id;
}
```

The cost should be minimal for most sensible implementations per http://en.cppreference.com/w/cpp/language/storage_duration#Static_local_variables

**Comment by LAJW** [comment]
Apart from the cost of function call there's going to be a performance hit similar to reading atomic ("magic static" - C++11 static variable initialisation is thread safe, so likely implemented using atomics). It's not implemented in VC2013 which we're supporting. If irep reference counting is not thread safe either, then that is a problem that can be safely ignored.

Moving `get_string_declaration` to header file and adding `inline` might alleviate @tautschnig's concern.

https://msdn.microsoft.com/en-us/library/hh567368.aspx

**Comment by reuk** [comment]
If we explicitly inline the function, wouldn't we end up with one copy of the static local variable for each translation unit?

**Comment by LAJW** [comment]
@reuk Apparently not.

> A static local variable in an extern inline function always refers to the same object.
> 7.1.2/4 - C++98/C++14 (n3797)

**Comment by reuk** [comment]
Some of the answers on StackOverflow say you may still end up with one instance per TU but the linker will pick a specific one to be the "canonical" instance, although I guess this may be implementation-dependent and may depend on LTO...

**Comment by tautschnig** [comment]
Don't get me wrong: I'm all for getting rid of the undefinedness that initialisation of static global objects induces; but this must not be merged until a performance evaluation is performed.

**Comment by reuk** [comment]
@tautschnig How would you propose to benchmark this PR?

**Comment by tautschnig** [comment]
This PR should not be benchmarked differently from any other PR, even though it is slightly easier here. Doing a summary profile across the regression tests might suffice for this PR if specific attention is paid to where in the profile the affected functions (`dstringt` and `dstringt::as_string`) show up.

Proper benchmarking would entail analysis of the effect on overall performance across a large set of benchmark inputs.

**Comment by reuk** [comment]
The following are times to run `make` from the `regression` directory. 

With patch:
```
make  177.83s user 11.76s system 92% cpu 3:25.03 total
```

Master:
```
make  177.36s user 11.84s system 93% cpu 3:21.84 total
```

**Comment by tautschnig** [comment]
That experiment won't provide more than the most basic sanity check. The bare minimum is actual profiling so that you can get information on what fraction of the total time is spent on the mentioned functions.

**Comment by martin-cs** [comment]
If it would help, I should be int he office tomorrow and have done quite a bit of profiling of CBMC and we could chat about it.

Also at one point @Degiorgio had an interest in this code...

**Comment by reuk** [comment]
Closing as duplicate of #1218 seeing as both are targeting develop now.

**Pull request closed without merging**